chips/rp2040/src/pwm.rs

```diff
@@ -0,0 +1,1254 @@
+                DivMode::Rising => CSR::DIVMOD::B_RISING,
+                DivMode::Falling => CSR::DIVMOD::B_FALLING,
+            });
+    }
+
+    // Set integral and fractional part of the clock divider
+    // RP 2040 uses a 8.4 fractional clock divider.
+    // The minimum value of the divider is   1 (int) +  0 / 16 (frac).
+    // The maximum value of the divider is 255 (int) + 15 / 16 (frac).
+    //
+    // **Note**: this method will do nothing if int == 0 || frac > 15.
+    fn set_divider_int_frac(&self, channel_number: ChannelNumber, int: u8, frac: u8) {
+        if int == 0 || frac > 15 {
+            return;
+        }
+        self.registers.ch[channel_number as usize]
+            .div
+            .modify(DIV::INT.val(int as u32));
+        self.registers.ch[channel_number as usize]
+            .div
+            .modify(DIV::FRAC.val(frac as u32));
+    }
+
+    // Set output pin A compare value
+    // If counter value < compare value A ==> pin A high
+    fn set_compare_value_a(&self, channel_number: ChannelNumber, cc_a: u16) {
+        self.registers.ch[channel_number as usize]
+            .cc
+            .modify(CC::A.val(cc_a as u32));
+    }
+
+    // Set output pin B compare value
+    // If counter value < compare value B ==> pin B high (if divmode == FreeRunning)
+    fn set_compare_value_b(&self, channel_number: ChannelNumber, cc_b: u16) {
+        self.registers.ch[channel_number as usize]
+            .cc
+            .modify(CC::B.val(cc_b as u32));
+    }
+
+    // Set compare values for both pins
+    fn set_compare_values_a_and_b(&self, channel_number: ChannelNumber, cc_a: u16, cc_b: u16) {
+        self.set_compare_value_a(channel_number, cc_a);
+        self.set_compare_value_b(channel_number, cc_b);
+    }
+
+    // Set counter top value
+    fn set_top(&self, channel_number: ChannelNumber, top: u16) {
+        self.registers.ch[channel_number as usize]
+            .top
+            .modify(TOP::TOP.val(top as u32));
+    }
+
+    // Get the current value of the counter
+    fn get_counter(&self, channel_number: ChannelNumber) -> u16 {
+        self.registers.ch[channel_number as usize]
+            .ctr
+            .read(CTR::CTR) as u16
+    }
+
+    // Set the value of the counter
+    fn set_counter(&self, channel_number: ChannelNumber, value: u16) {
+        self.registers.ch[channel_number as usize]
+            .ctr
+            .modify(CTR::CTR.val(value as u32));
+    }
+
+    // Increments the value of the counter
+    //
+    // The counter must be running at less than full speed. The method will return
+    // once the increment is complete.
+    fn advance_count(&self, channel_number: ChannelNumber) {
+        self.registers.ch[channel_number as usize]
+            .csr
+            .modify(CSR::PH_ADV::SET);
+        while self.registers.ch[channel_number as usize]
+            .csr
+            .read(CSR::PH_ADV)
+            == 1
+        {}
```

Comment:
Can you comment this with the maximum number of cycles you would expect this to block?

Reply:
Additionally I suggest defining a `wait_for` function as in https://github.com/OxidosAutomotive/tock/blob/026361c7e830586c3eb25f2345b0449618c16078/chips/stm32f4xx/src/can.rs#L502-L517 and suing it like https://github.com/OxidosAutomotive/tock/blob/026361c7e830586c3eb25f2345b0449618c16078/chips/stm32f4xx/src/can.rs#L532-L534 and report an error.

Reply:
The datasheet doesn't say anything about the maximum number of cycles the pooling should be done. Should I assume that the pooling should be over after 1 PWM period?

Reply:
Just wait a few cycles (maybe max 100) before reporting an error.

Reply:
Implemented `wait_for()` method.